addons/loyalty/models/loyalty_card.py

```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from uuid import uuid4
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+class LoyaltyCard(models.Model):
+    _name = 'loyalty.card'
+    _inherit = ['mail.thread']
+    _description = 'loyalty Coupon'
+    _rec_name = 'code'
+
+    # This model does not only represent coupons but also wallets and loyalty cards
+
+    @api.model
+    def _generate_code(self):
+        """
+        Barcode identifiable codes.
+        """
+        return '044' + str(uuid4())[4:-8][3:]
+
+    program_id = fields.Many2one('loyalty.program')
```

Comment:
?
```suggestion
    program_id = fields.Many2one('loyalty.program', ondelete='cascade')
```

Reply:
Or restrict actually.

Reply:
We could cascade but `coupon_id` is `ondelete='restrict'` on `sale.order.line` so if it is used, it will be restricted.
Knowing that isn't it better to go for restrict here aswell? And archive programs instead?